library/Console-posix.cpp

```diff
@@ -542,26 +546,17 @@ namespace DFHack
                 }
                 lock->lock();
                 const int old_cursor = raw_cursor;
-                /* Only autocomplete when the callback is set. It returns < 0 when
-                 * there was an error reading from fd. Otherwise it will return the
-                 * character that should be handled next. */
+
                 if (c == 9)
                 {
-                    /*
-                    if( completionCallback != NULL) {
-                        c = completeLine(fd,prompt,buf,buflen,&len,&pos,cols);
-                        // Return on errors
-                        if (c < 0) return len;
-                        // Read next character when 0
-                        if (c == 0) continue;
-                    }
-                    else
+                    std::string part = toLocaleMB(raw_buffer);
+                    auto command = command_list.lower_bound(part);
+                    if(command != command_list.end() && command->substr(0, part.size()) == part)
                     {
-                        // ignore tab
-                        continue;
+                        raw_buffer = u32string(command->begin(), command->end());
+                        raw_cursor = command->size();
```

Comment:
Have you checked what happens if the cursor isn't originally at the end of the line when tab is pressed? (I'm not entirely sure why you'd do that, so weird behavior is fine with me for now, as long as it doesn't crash)

Reply:
It does not crash. If cursor is in the middle of command it jumps to the end if command is a mach. If it is not a mach it just stays where it was